sourcegraph/sourcegraph
# Clarify Perforce p4.passwd / ticket format

- Fixes https://github.com/sourcegraph/sourcegraph/issues/43307

In short, the term `ticket` in ambiguous in Perforce, and the Perforce CLI only accepts one of the two things it means (it accepts the shorter of the two). We want to make sure that users enter the shorter one.
Conveniently, the shorter one can _never_ contain a **colon**, while the longer one _always_ contains one.

So, three changes here:
1. Added more instructions to the tooltip, to guide the user _before_ they save their settings.
2. Added a colon-based validation to avoid _saving_ the wrong setting.
3. Added a test for the new validation

Note: Unfortunately, I the tooltip wasn't displayed locally because of front-end error unrelated to my changes (see screenshot ↓), but I've checked our other tooltips, and we have some that have similar lengths, and that contain backticks, so this should be fine.

## Test plan

I've tested it with this process:
- Go to Site admin | Manage code hosts | Add code host
- See tooltip (couldn't actually see how it looks, see screenshot)
- Set `p4.passwd` to a value that contains a colon
- Click "Add repositories"
- Error message should come up (see screenshot)
- Set `p4.passwd` to a value that doesn't contain a colon
- It should save.

![CleanShot 2022-11-10 at 12 45 07](https://user-images.githubusercontent.com/2552265/201082757-31d1bf4d-12ff-46ba-b2ca-4d51703da0a3.png)
